src/raw/serde/seeded_visitor.rs

```diff
@@ -0,0 +1,564 @@
+use std::{borrow::Cow, convert::TryFrom, fmt::Formatter, ops::Range};
+
+use serde::{
+    de::{DeserializeSeed, Error as SerdeError, MapAccess, SeqAccess, Visitor},
+    Deserializer,
+};
+use serde_bytes::ByteBuf;
+
+use crate::{
+    de::MIN_BSON_DOCUMENT_SIZE,
+    extjson::models::{
+        BorrowedBinaryBody,
+        BorrowedDbPointerBody,
+        BorrowedRegexBody,
+        TimestampBody,
+    },
+    oid::ObjectId,
+    raw::{RAW_ARRAY_NEWTYPE, RAW_BSON_NEWTYPE, RAW_DOCUMENT_NEWTYPE},
+    spec::{BinarySubtype, ElementType},
+    RawDocumentBuf,
+};
+
+use super::CowStr;
+
+/// A copy-on-write byte buffer containing raw BSON bytes. The inner value starts as `None` and
+/// transitions to either `Cow::Borrowed` or `Cow::Owned` depending upon the data visited.
```

Comment:
Is `CowByteBuffer` needed? Consider using a non-copy-on-write buffer it it helps simplify.

IIUC the entry points to `SeededVisitor` are `iterate_map` and `visit_seq`. Both call `append_cstring` which I expect will copy the buffer.

Reply:
I forgot to mention yesterday that `visit_borrowed_bytes` is also a valid entry point for this method when a slice of document bytes is provided, which will store a reference to the bytes in `Cow::Borrowed`.